### DISCUSSION TITLE
i3 module continually crashes shortly after launch

### DISCUSSION THREAD

**Discussion by anoduck** [polybar/polybar#2941] [Q&A] [answered]
I was looking for a way to achieve a working setup with i3. The module for i3 will not stay up for long after launch, and as a result I am constantly stopping polybar and relaunching it. 

The output I have caught so far is:

```
Segmentation fault (core dumped)  

polybar|error: module/i3: Failed to connect to i3: errno 2 (No such file or directory)

polybar|notice:  Stopping 'module/i3'...
```

Another probably connected error message is:

```
polybar|warn:  module/i3: Attempting to reconnect socket (reason: Unexpected EOF while reading header)
```
I also noticed polybar was attempting to reconnect to a socket with a different name than those that were available. 

I am unsure if there is another ipc client I could try to implement or a different configuration variable that could be used that is not listed in wiki.

**Comment by patrick96** [comment]
Very much looks like a bug. Are you doing anything when the i3 module crashes?

> I also noticed polybar was attempting to reconnect to a socket with a different name than those that were available.

Where did you see that?

Do the first three lines of output you showed appear in that order with nothing in between? Because that is very weird behavior, I would expect polybar to stop running after a Segmentation fault and  nothing to be logged.

Also, what versions of polybar and i3 are you running?

**Reply by anoduck** [reply]
> Very much looks like a bug. Are you doing anything when the i3 module crashes?

Nothing particularly noticeable. I will just look down and notice the entire section that lists workspaces will be missing from the bar. 
 
> > I also noticed polybar was attempting to reconnect to a socket with a different name than those that were available.
> 
> Where did you see that?

This was an error in observation on my part. At the time of writing this, I was unaware polybar created an additional socket for legacy ipc in `/tmp`, and was confusing the two. 

Oddly, all those open legacy sockets are not cleared when I stop the script. 

```
polybar-1000/          polybar_mqueue.14645|  polybar_mqueue.23828|  polybar_mqueue.31394|  polybar_mqueue.45402|  polybar_mqueue.50490|  polybar_mqueue.53597|  polybar_mqueue.60463|  polybar_mqueue.76235|  polybar_mqueue.78945|  polybar_mqueue.97880|
polybar1.log           polybar_mqueue.19291|  polybar_mqueue.2611|   polybar_mqueue.34413|  polybar_mqueue.46239|  polybar_mqueue.53311|  polybar_mqueue.55824|  polybar_mqueue.68952|  polybar_mqueue.76635|  polybar_mqueue.82|     polybar_mqueue.99910|
polybar2.log           polybar_mqueue.22340|  polybar_mqueue.28519|  polybar_mqueue.35897|  polybar_mqueue.46976|  polybar_mqueue.53455|  polybar_mqueue.57756|  polybar_mqueue.7312|   polybar_mqueue.7890|   polybar_mqueue.91213|
```

> 
> Do the first three lines of output you showed appear in that order with nothing in between? Because that is very weird behavior, I would expect polybar to stop running after a Segmentation fault and nothing to be logged.
> 

Weird indeed, and it is confirmed to be exactly in that order in the log file.

```
 485   │ polybar|warn:  Dropping unmatched character 💤 (U+1f4a4) in 'folke/lazy.nvim: 💤 A modern plugin manager for Neo...'
 486   │ polybar|warn:  Dropping unmatched character 💤 (U+1f4a4) in 'folke/lazy.nvim: 💤 A modern plugin manager for Neo...'
 487   │ Segmentation fault (core dumped) 
 488   │ polybar|error: module/i3: Failed to connect to i3: errno 2 (No such file or directory)
 489   │ polybar|notice:  Stopping 'module/i3'...
 490   │ polybar|warn:  Dropping unmatched character 💤 (U+1f4a4) in 'folke/lazy.nvim: 💤 A modern plugin manager for Neo...'
 491   │ polybar|warn:  Dropping unmatched character 💤 (U+1f4a4) in 'folke/lazy.nvim: 💤 A modern plugin manager for Neo...'
 492   │ polybar|warn:  Dropping unmatched character 💤 (U+1f4a4) in 'folke/lazy.nvim: 💤 A modern plugin manager for Neo...'
```

> Also, what versions of polybar and i3 are you running?

Polybar: 3.6.3 with -alsa +curl +i3 +mpd -network(wireless-tools) +pulseaudio +xkeyboard
i3: 4.22

------

Apparently, polybar is the defacto alternative for a status bar in i3. I looked to see if there was an alternative, and didn't find anything I liked. Also, considered moving to bspwm, but after thumbing through the configuration files, I didn't feel the love.

**Reply by patrick96** [reply]
I have a suspicion  what's happening here. When polybar connects to i3, it runs the command `i3 --get-socketpath` to get the location of i3's IPC socket.
I think that command crashes and produces the "Segmentation fault (core dumped)" message in the log and as a result polybar can't connect to i3 and the module crashes.

In that case, this would be an i3 bug. Though polybar could do a better job of error reporting.

Are you comfortable building polybar from source (e.g. from a PR)? I will prepare a patch that will give better error messages and should also confirm or refute my suspicion.

**Reply by patrick96** [reply]
Alright, I created #2947, which should tell us more clearly what's going wrong. If you can, please try it out and report the output you get when the i3 module fails again.

**Reply by anoduck** [reply]
@patrick96 I like your theory, it is a good explanation of the errors observed. 

Although, I had previously built it before noticing listed in the ports tree, I do not remember it throwing the `sys/inotify.h file not found` error. I have a sneaky suspicion it is looking for `inotify.h` in the wrong location, because it is located at `/usr/local/include/inotify/sys/inotify/inotify.h`.

In a standard build operation, I vaguely remember I can drop an `--include` flag informing configure where the missing library is located, but it is a little fuzzy how to do this with cmake. Either that, or my brain is a little tired after staring at antenna designs.

Ok, I am going to have to tell cmake where it is located before cmake generates the build files. It appears cmake misconfigured where the system's includes directory is, setting it to `/usr/include`

From `CMakeFiles/CMakeConfigureLog.yaml`:

```bash
#include "..." search starts here:
#include <...> search starts here:
  /usr/lib/clang/13.0.0/include
  /usr/include
Parsed C implicit include dir info: rv=done
  found start of include info
  found start of implicit include info
    add: [/usr/lib/clang/13.0.0/include]
    add: [/usr/include]
  collapse include dir [/usr/lib/clang/13.0.0/include] ==> [/usr/lib/clang/13.0.0/include]
  collapse include dir [/usr/include] ==> [/usr/include]
  implicit include dirs: [/usr/lib/clang/13.0.0/include;/usr/include]
```

And again, listed here in the output from running `cmake -DCMAKE_CXX_COMPILER="clang++"`:

```bash
--  Build:
--    Version: 3.6.3-181-g3984c2e0-dev
--    Type: Release
--    CXX: /usr/bin/clang++  -O3 -DNDEBUG -Wall -Wextra -Wpedantic -Wdeprecated-copy-dtor -Wsuggest-override -Wno-error=parentheses-equality -Wno-zero-length-array
--    LD: /usr/bin/ld.lld  -Wall -Wextra -Wpedantic -Wdeprecated-copy-dtor -Wsuggest-override -Wno-error=parentheses-equality -Wno-zero-length-array  -Wall -Wextra -Wpedantic -Wdeprecated-copy-dtor -Wsuggest-override -Wno-error=parentheses-equality -Wno-zero-length-array
--    sphinx-build: /usr/local/bin/sphinx-build 
--  Install Paths:
--    PREFIX:     /usr  # <--- Should be: /usr/local
--    BINDIR:     /usr/bin  # <--- Should be: /usr/local/bin
--    DATADIR:    /usr/share  # <--- Should be: /usr/local/share
--    DOCDIR:     /usr/share/doc/polybar # <--- Ditto, wrong prefix
--    MANDIR:     /usr/man  # <--- Should be: /usr/local/man
--    SYSCONFDIR: /etc
--  Targets:
-- [X]   polybar
-- [X]   polybar-msg
-- [ ]   testsuite
-- [X]   documentation
-- [X]      html
-- [X]      man
-- [X]   default config
-- [X]   shell files
--  Module support:
-- [ ]   alsa
-- [X]   curl (8.0.1)
-- [X]   i3
-- [X]   mpd (2.20)
-- [ ]   network (wireless-tools)
-- [X]   pulseaudio (16.1)
-- [X]   xkeyboard (1.15)
--  X extensions:
-- [X]   xcb-randr (1.15)
-- [X]   xcb-randr (monitor support) (1.15)
-- [X]   xcb-composite (1.15)
-- [X]   xcb-xkb (1.15)
-- [X]   xcb-xrm (1.0)
-- [X]   xcb-cursor (0.1.4)
--  Log options:
-- [X]   Trace logging
-- Configuring done (10.1s)
-- Generating done (0.3s)
-- Build files have been written to: /home/user/polybar/build
```

I will have to work on correcting the prefix, and run it again tomorrow.

**Reply by patrick96** [reply]
I have never seen a system that has its system header files in `/usr/local/include` and not `/usr/include`. Also, both gcc and clang on my system also by default always search in `/usr/local/include`.
What distro are you running?

In any case, you can run `cmake -DCMAKE_CXX_COMPILER="clang++" -DCMAKE_INSTALL_PREFIX=/usr/local ..` to tell polybar to install to `/usr/local`.

Maybe this also fixes the include issue, but I'm skeptical.
The `inotify.h` file is includes as `#include <sys/inotify.h>`, so its path must end with `/sys/inotify.h` and on your system the path ends with `/inotify/inotify.h`.

**Comment by anoduck** [comment]
I apologize, I copied down the location of inotify incorrectly. 

Here is the output of `locate inotify.h`:

```bash
$\Dir > polybar/build > on i3-socket-errors [!] ►⦙ locate inotify.h
/home/user/polybar/include/utils/inotify.hpp
/usr/local/include/inotify/sys/inotify.h
/usr/local/include/wx-3.1/wx/unix/fswatcher_inotify.h
/usr/local/include/wx-3.2/wx/unix/fswatcher_inotify.h
/usr/local/share/ports-readme/path/devel/libinotify.html
/usr/ports/pobj/iridium-2023.03.111/iridium-browser-2023.03.111/base/files/file_path_watcher_inotify.h
/usr/ports/pobj/polybar-3.6.3/polybar-3.6.3/include/utils/inotify.hpp
```

I was able to correctly configure the prefix and sysconfig settings with:

```bash
cmake -DCMAKE_CXX_COMPILER="clang++" -DCMAKE_INSTALL_PREFIX="/usr/local" -DCMAKE_INSTALL_SYSCONFDIR="/etc" ..
```

Unfortunately, it still could not find `/sys/inotify.h`

**Reply by patrick96** [reply]
Try this:

```
cmake -DCMAKE_CXX_COMPILER="clang++" -DCMAKE_INSTALL_PREFIX="/usr/local" -DCMAKE_INSTALL_SYSCONFDIR="/etc" -DCMAKE_CXX_FLAGS="-I/usr/local/include" ..
```

**Reply by patrick96** [reply]
Hmm, since the inotify header is at `/usr/local/include/inotify/sys/inotify.h`, maybe it should actually be

```
cmake -DCMAKE_CXX_COMPILER="clang++" -DCMAKE_INSTALL_PREFIX="/usr/local" -DCMAKE_INSTALL_SYSCONFDIR="/etc" -DCMAKE_CXX_FLAGS="-I/usr/local/include -I/usr/local/include/inotify" ..
```

Is this a BSD system?

**Reply by anoduck** [reply]
> Is this a BSD System?

Yes, It is “OpenBSD CURRENT SNAPSHOT 7.3 amd64”, and yes, you are a brilliant programmer. OpenBSD is just odd like that…

----

I ended up splitting the last flag for the latter command in twain. So, what I used was:

```bash
cmake -DCMAKE_CXX_COMPILER="clang++" -DCMAKE_INSTALL_PREFIX="/usr/local" -DCMAKE_INSTALL_SYSCONFDIR="/etc" -DCMAKE_CXX_FLAGS="-I/usr/local/include" -DCMAKE_CXX_FLAGS="-I/usr/local/include/inotify"  ..
```

This appeared to build well, until the gods chose to frown upon my fortunes. 

```bash
ld: error: unable to find library -lcairo
ld: error: unable to find library -lfontconfig
ld: error: unable to find library -lfreetype
ld: error: unable to find library -luv
ld: error: unable to find library -ljsoncpp
ld: error: unable to find library -lcurl
ld: error: unable to find library -lmpdclient
ld: error: unable to find library -lpulse
ld: error: unable to find library -lxcb-randr
ld: error: unable to find library -lxcb-composite
ld: error: unable to find library -lxcb-xkb
ld: error: unable to find library -lxcb-cursor
ld: error: unable to find library -lxcb-xrm
ld: error: unable to find library -lxcb
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
*** Error 1 in . (bin/CMakeFiles/polybar.dir/build.make:108 'bin/polybar': cd /home/user/polybar/build/bin && /usr/local/bin/cm...)
*** Error 2 in . (CMakeFiles/Makefile2:575 'bin/CMakeFiles/polybar.dir/all': make -s -f bin/CMakeFiles/polybar.dir/build.make bin/CMakeFiles...)
*** Error 2 in /home/user/polybar/build (Makefile:136 'all': make -s -f CMakeFiles/Makefile2 all)
```

These libraries were found during the execution of cmake. Why they are now lost, when once they were found, I am unsure of.

Here is output from cmake, confirming their existence and location:

```bash
Found Threads: TRUE  
-- Checking for module 'cairo-fc'
--   Found cairo-fc, version 1.17.8
-- Found CairoFC: /usr/local/include;/usr/local/include/cairo;/usr/local/include/libpng16;/usr/X11R6/include;/usr/X11R6/include/pixman-1;/usr/X11R6/include/freetype2 (found version "1.17.8") 
-- Checking for module 'libuv'
--   Found libuv, version 1.44.2
-- Found LibUV: /usr/local/include (found suitable version "1.44.2", minimum required is "1.3.0") 
-- Checking for module 'libcurl'
--   Found libcurl, version 8.0.1
-- Found CURL: /usr/local/include (found version "8.0.1") 
-- Checking for module 'libmpdclient'
--   Found libmpdclient, version 2.20
-- Found LibMPDClient: /usr/local/include (found version "2.20") 
-- Checking for module 'libpulse'
--   Found libpulse, version 16.1
-- Found LibPulse: /usr/local/include (found version "16.1") 
-- Checking for module 'xcb-randr'
--   Found xcb-randr, version 1.15
-- Found Xcb_RANDR: /usr/X11R6/include (found suitable version "1.15", minimum required is "1.12") 
-- Checking for module 'xcb-composite'
--   Found xcb-composite, version 1.15
-- Found Xcb_COMPOSITE: /usr/X11R6/include (found version "1.15") 
-- Checking for module 'xcb-xkb'
--   Found xcb-xkb, version 1.15
-- Found Xcb_XKB: /usr/X11R6/include (found version "1.15") 
-- Checking for module 'xcb-cursor'
--   Found xcb-cursor, version 0.1.4
-- Found Xcb_CURSOR: /usr/X11R6/include (found version "0.1.4") 
-- Checking for module 'xcb-xrm'
--   Found xcb-xrm, version 1.0
-- Found Xcb_XRM: /usr/X11R6/include (found version "1.0") 
-- The C compiler identification is Clang 13.0.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Checking for module 'xcb-proto'
--   Found xcb-proto, version 1.15.2
-- Found PythonInterp: /usr/local/bin/python3 (found suitable version "3.10.11", minimum required is "3.5") 
-- XCB[XCB]: Found component XCB
-- Found XCB_XCB: /usr/X11R6/lib/libxcb.so.4.1  
-- XCB[ICCCM]: Found component ICCCM
-- Found XCB_ICCCM: /usr/X11R6/lib/libxcb-icccm.so.4.0  
-- XCB[EWMH]: Found component EWMH
-- Found XCB_EWMH: /usr/X11R6/lib/libxcb-ewmh.so.2.0  
-- XCB[UTIL]: Found component UTIL
-- Found XCB_UTIL: /usr/X11R6/lib/libxcb-util.so.0.0  
-- XCB[IMAGE]: Found component IMAGE
-- Found XCB_IMAGE: /usr/X11R6/lib/libxcb-image.so.2.0  
-- Found XCB: /usr/X11R6/lib/libxcb.so.4.1;/usr/X11R6/lib/libxcb-icccm.so.4.0;/usr/X11R6/lib/libxcb-ewmh.so.2.0;/usr/X11R6/lib/libxcb-util.so.0.0;/usr/X11R6/lib/libxcb-image.so.2.0  
-- Searching for xcbgen with /usr/local/bin/python3
-- Found xcbgen in /usr/local/lib/python3.10/site-packages
-- XCB[RANDR]: Found component RANDR
-- Found XCB_RANDR: /usr/X11R6/lib/libxcb-randr.so.2.3  
-- Found XCB: /usr/X11R6/lib/libxcb.so.4.1;/usr/X11R6/lib/libxcb-icccm.so.4.0;/usr/X11R6/lib/libxcb-ewmh.so.2.0;/usr/X11R6/lib/libxcb-util.so.0.0;/usr/X11R6/lib/libxcb-image.so.2.0;/usr/X11R6/lib/libxcb-randr.so.2.3  
-- XCB[COMPOSITE]: Found component COMPOSITE
-- Found XCB_COMPOSITE: /usr/X11R6/lib/libxcb-composite.so.1.0  
-- Found XCB: /usr/X11R6/lib/libxcb.so.4.1;/usr/X11R6/lib/libxcb-icccm.so.4.0;/usr/X11R6/lib/libxcb-ewmh.so.2.0;/usr/X11R6/lib/libxcb-util.so.0.0;/usr/X11R6/lib/libxcb-image.so.2.0;/usr/X11R6/lib/libxcb-randr.so.2.3;/usr/X11R6/lib/libxcb-composite.so.1.0  
-- XCB[XKB]: Found component XKB
-- Found XCB_XKB: /usr/X11R6/lib/libxcb-xkb.so.0.1  
-- Found XCB: /usr/X11R6/lib/libxcb.so.4.1;/usr/X11R6/lib/libxcb-icccm.so.4.0;/usr/X11R6/lib/libxcb-ewmh.so.2.0;/usr/X11R6/lib/libxcb-util.so.0.0;/usr/X11R6/lib/libxcb-image.so.2.0;/usr/X11R6/lib/libxcb-randr.so.2.3;/usr/X11R6/lib/libxcb-composite.so.1.0;/usr/X11R6/lib/libxcb-xkb.so.0.1  
-- xpp: including xcb proto composite.xml
-- xpp: including xcb proto randr.xml
-- xpp: including xcb proto xkb.xml
-- xpp: including xcb proto xproto.xml
```

I tried rebuilding the ldd cache, and did an excellent job buggering it up. Thankfully, a reboot put me back in business.

So, this is getting unnecessarily difficult. Regardless, I am figuring out why those library files were not found.

**Reply by anoduck** [reply, marked as answer]
I believe I have found a resolution, it is mentioned in the now closed [issue report](https://github.com/polybar/polybar/issues/2942#issuecomment-1872903950).
